Add MCO overview docs page (#51)
GoogleCloudPlatform/gke-fleet-management: multi-cluster-orchestrator/docs/overview.md

# Overview of Multi-Cluster Orchestrator

Multi-Cluster Orchestrator (MCO) dynamically schedules an application’s

> multi-cluster and multi-region deployment, allowing the application to

Author: Right. I think people will understand what we mean, but if you think it's unclear I'm open to suggestions

automatically scale out to more clusters or scale in to fewer clusters as
needed.

MCO schedules an application workload onto clusters based on:

> *   The preferences specified by the user, such which clusters are eligible to

Author: Fixed

    run the workload and which clusters/regions are more desirable to use
*   Information about region capacity, such as which regions currently have
    certain machine types available
*   Current load on the application

These scheduling decisions are consumed by the CD tool (a.k.a. workload delivery
system) which provisions the workload onto clusters or removes the workload from
clusters as required.

> MCO requires a “hub cluster” which contains metadata about the other clusters in

Author: Done

the fleet (in the form of ClusterProfile resources). The hub cluster is used to

> run special workloads that manage the rest of the fleet such as MCO itself or CD

Author: Done

> ClusterProfile resources). The hub cluster is used to run multicluster

Reviewer: suggestion

```suggestion
[ClusterProfile](https://multicluster.sigs.k8s.io/concepts/cluster-profile-api/) resources). The hub cluster is used to run multicluster
```

The application workload is deployed to other clusters in the fleet. MCO can be
used along with a multi-cluster load balancer such as a Google Cloud
[multi-cluster
Gateway](https://cloud.google.com/kubernetes-engine/docs/how-to/deploying-multi-cluster-gateways)
load balancer to dynamically route incoming traffic to the various workload
clusters.

# Architecture

## Design Principles

Other multi-cluster scheduling efforts have tried to do a lot of things all at
once. MCO has intentionally tried to avoid this and instead focuses on solving
the relatively narrow problem of selecting *which* clusters should be running
the workload at a given point in time (including scaling to more/fewer
clusters).

Closely related to this, one of the core design principles of MCO is that the
MCO controller does not interact directly with the clusters that it is managing
except for consuming metrics which they publish. This means that MCO does not
directly provision workloads on clusters or scale the size of the workload on
each cluster.

Instead, the delivery of workloads to clusters is decoupled from the core of
MCO. This allows workload delivery to be pluggable such that it can be

> integrated/delegated to existing systems such as ArgoCD or Flux (which the user

Reviewer: ```suggestion
integrated/delegated to existing systems such as Argo CD or Flux (which the user
```

runs in the same hub cluster as the MCO controller).

Locally scaling the workload within each cluster is left to the standard

> Kubernetes HorizontalPodAutoscaler.

Reviewer: suggestion

```suggestion
Kubernetes [HorizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/).
```

## Architecture Diagram

This diagram shows the high-level architecture of Multi-Cluster Orchestrator
when running on Google Cloud. The various components are explained below.

![Diagram of MCO Archtecture](images/mco_architecture.svg)

## Placement API

The core of Multi-Cluster Orchestrator is the Placement API. The Placement API
is a CRD which defines a workload to be scheduled and autoscaled onto multiple
clusters.

The spec of the Placement API defines which clusters are eligible to run the
workload and defines the autoscaling parameters for the workload (see below for
more details about autoscaling).

The status of the Placement API contains the set of clusters where the workload
is currently scheduled. This is consumed by the workload delivery system.

Here’s an example of an autoscaling Placement. The clusters that are eligible
to run the workload are defined by the rules section. We start with all of the
clusters for which ClusterProfiles exist, then we filter that using a regex
matching the ClusterProfile’s namespace/name.

```
apiVersion: orchestra.multicluster.x-k8s.io/v1alpha1
kind: MultiKubernetesClusterPlacement
metadata:

>   name: cluster-list-placement

Author: Good catch, fixed

Initially, we have a set of eligible clusters as defined by the rules in the
Placement. The MCO Controller will start by scaling out the workload to the
first eligible cluster — adding that cluster to the Placement in the ACTIVE
state.

If a current cluster becomes ineligible to run the workload, it transitions to
the EVICTING state and the workload is drained and removed from it.

Part of the MCO Controller is the autoscaler which makes scaling decisions based
on metrics from the clusters (see the Autoscaling section below for details).

If the autoscaler decides that fewer clusters are needed, it moves one of the
ACTIVE clusters into the SCALING_IN state and we start draining the workload
from the cluster. Once the workload gets down to one remaining pod, it is
removed from the cluster entirely.

If the autoscaler decides that more clusters are needed, it first checks if
there is a SCALING_IN cluster available and if so moves it back to ACTIVE.
Otherwise it picks the next eligible cluster and adds it to the set of ACTIVE
clusters.

Health checking is a planned feature but is not currently implemented.

> Note: Proactively draining pods from clusters is coming soon but not currently

Reviewer: If you want an "alert" rather than plain text note: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

Author: Nice, TIL. Done.

available. This means that clusters will only be removed when they have
naturally scaled down to very few pods.

## Workload Delivery

As mentioned above, workload delivery is decoupled from the core MCO Controller
which allows for multiple implementations.

Each implementation should watch the status field of each Placement and ensure
that the workload is provisioned on clusters when they are added to the status
field and removed from clusters when they are removed from the status field.

The only workload delivery system that is currently available is an Argo CD
[ApplicationSet](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/)
[generator](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators/)
[plugin](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-Plugin/)

> which allows users to dynamically target Argo CD Applications to clusters.

Author: Added

cluster profiles for all clusters in the fleet can be [automatically generated
in the hub cluster by a syncer running in Google
Cloud](https://cloud.google.com/kubernetes-engine/fleet-management/docs/create-inventory).
When a cluster is added or removed from the fleet, the list of cluster profiles
is updated automatically by the syncer.

Using the syncer is optional; the cluster profiles could also be created and
managed manually.

# Autoscaling

The autoscaler is part of the MCO Controller.

**In the current implementation, the autoscaler tries to maintain a target
number of clusters which are below their capacity ceiling — i.e. clusters on
which more pods can still be scheduled.** This will likely evolve in the future
to be more flexible/sophisticated.

Metrics published by each cluster are used to determine which clusters are
currently at their capacity ceiling. To accomplish this, it uses the following
query logic:

1.  Check if the cluster is at its capacity ceiling due to the current nodes
    being at capacity and the cluster-autoscaler being unable to obtain more
    nodes.

    To check this, we check whether: a) there are pending pods for the workload

>     and b) there have been FailedScaleUp events from the cluster-autoscaler (the

Reviewer: If you want to use code formatting for these and similar:

```suggestion
    and b) there have been `FailedScaleUp` events from the cluster-autoscaler (the
```
